pkg/importer/dataStream_test.go

```diff
@@ -7,416 +7,409 @@ import (
+var testfiles = createTestData()
+var stringRdr = strings.NewReader("test data for reader 1")
+var rdrfile, _ = os.Open(filepath.Join(imageDir, "cirros-qcow2.img"))
+var testFileRdrs = bufio.NewReader(rdrfile)
+
 type fakeQEMUOperations struct {
 	e1 error
 	e2 error
 	e3 error
 }
 
-func (o *fakeQEMUOperations) ConvertQcow2ToRaw(string, string) error {
-	return o.e1
-}
+var _ = Describe("Data Stream", func() {
+	var ts *httptest.Server
 
-func (o *fakeQEMUOperations) ConvertQcow2ToRawStream(*url.URL, string) error {
-	return o.e2
-}
+	BeforeEach(func() {
+		By("[BeforeEach] Creating test server")
+		ts = createTestServer(imageDir)
+	})
 
-func (o *fakeQEMUOperations) Validate(string, string) error {
-	return o.e3
-}
+	AfterEach(func() {
+		By("[AfterEach] closing test server")
+		ts.Close()
+	})
 
-func NewFakeQEMUOperations(e1, e2, e3 error) image.QEMUOperations {
-	return &fakeQEMUOperations{e1, e2, e3}
-}
+	table.DescribeTable("with import source should", func(image, accessKeyID, secretKey string, qemu bool, want []byte, wantErr bool) {
+		if image != "" {
+			image = ts.URL + "/" + image
+		}
+		By("Creating datastream")
+		ds, err := NewDataStream(image, accessKeyID, secretKey)
+		if !wantErr {
+			Expect(err).NotTo(HaveOccurred())
+			resultBuffer := make([]byte, len(want))
+			By("Reading data from buffer")
+			ds.Read(resultBuffer)
+			Expect(ds.accessKeyID).To(Equal(accessKeyID))
+			Expect(ds.secretKey).To(Equal(secretKey))
+			Expect(ds.qemu).To(Equal(qemu))
+			Expect(reflect.DeepEqual(resultBuffer, want)).To(BeTrue())
+		} else {
+			Expect(err).To(HaveOccurred())
+		}
+	},
+		table.Entry("expect NewDataStream to succeed with valid image", "cirros-qcow2.img", "", "", true, cirrosData, false),
+		table.Entry("expect NewDataStream to fail with non existing image", "badimage.iso", "", "", false, nil, true),
+		table.Entry("expect NewDataStream to fail with invalid or missing image", "", "", "", false, nil, true),
+		table.Entry("expect NewDataStream to succeed with valid iso image", "tinyCore.iso", "accessKey", "secretKey", false, tinyCoreData, false),
+	)
+
+	It("can close all readers", func() {
+		ds, err := NewDataStream(ts.URL+"/tinyCore.iso", "", "")
+		Expect(err).NotTo(HaveOccurred())
+		emptyBuffer(ds)
+		By("Closing data stream")
+		err = ds.Close()
+		Expect(err).NotTo(HaveOccurred())
+	})
 
-func NewQEMUAllErrors() image.QEMUOperations {
-	err := errors.New("qemu should not be called from this test override with replaceQEMUOperations")
-	return NewFakeQEMUOperations(err, err, err)
-}
+	table.DescribeTable("can build selector", func(image, ep string, wantErr bool) {
+		if ep == "" {
+			ep = ts.URL
+		}
+		ds, err := NewDataStream(ep+"/"+image, "", "")
+		if !wantErr {
+			Expect(err).NotTo(HaveOccurred())
+			defer ds.Close()
+			emptyBuffer(ds)
+			err = ds.dataStreamSelector()
+			Expect(err).NotTo(HaveOccurred())
+		} else {
+			Expect(err).To(HaveOccurred())
+		}
+	},
```

Comment:
Should there be a check that the correct scheme was returned? I guess for now it's always http but maybe later s3 will be added?

Reply:
Unless we can somehow fake an S3 server, I think its always going to be http, and thus is not very interesting to test for (since we add ts.URL to the ep if its empty)

Reply:
It looks like minio has a mock s3 server so this could happen...

Reply:
We are still investigating how we can do S3 tests, I don't think the original tests that I copied checked for scheme either.

Reply:
@awels you are correct.